1-Draft/RFC00XX-Native-Command-Error-Handling.md

```diff
@@ -0,0 +1,334 @@
+benefit when expecting non-terminating errors in normal execution such as non-responsive computers
+from a list. This default behavior is controlled with the preference variable
+`$ErrorActionPreference` default of `Continue`.
+
+In production, often customers prefer that script execution stops when a non-terminating error
+occurs. This is particularly true in CI where the preference is to fail fast. PowerShell currently
+supports customers with this ability by setting the `$ErrorActionPreference` variable in the script
+to `Stop`.
+
+```Powershell
+$ErrorActionPreference = 'Stop'
+```
+
+Native commands usually return an exit code to the calling application which will be zero for
+success or non-zero for failure. However, native commands currently do not participate in the
+PowerShell error stream. Redirected `stderr` output is not interpreted the same as the PowerShell
+error stream as many native commands use `stderr` as information/verbose stream and thus only the
+exit code matters. Users working with native commands in their scripts will need to check the
+execution status after each call using a helper function similar to below:
+
+```Powershell
+if ($LASTEXITCODE -ne 0)
+{
+    throw "Command failed. See above errors for details"
+}
+ ```
+
+Simply relaying the errors through the error stream isn't the solution. The example itself doesn't
+support all cases as `$?` can be false from a cmdlet or function error, making `$LASTEXITCODE`
+stale.
+
+In POSIX shells, this need to terminate on command error is addressed by the `set -e` configuration,
+which causes the script to exit when a command fails. In addition, to ensure that an error is
+returned if any command in a pipeline fails, POSIX shells address this need with `set -o pipefail`
+configuration.
+
+This specification proposes a similar idea, but adapted to the PowerShell conventions of preference
+variables and catchable, self-describing, terminating error objects. This proposal adds the
+equivalent functionality of `set -eo pipefail` to stop execution and return an error if any command in a pipeline
+fails.
+
+ The specification and alternative proposals are based on the
+ [Equivalent of bash `set -e` #3415](https://github.com/PowerShell/PowerShell/issues/3415)
+ committee review of the associated
+ [pull request](https://github.com/PowerShell/PowerShell/pull/3523), and
+ [implementation plan](https://github.com/PowerShell/PowerShell-RFC/pull/88#issuecomment-613653678)
+
+## Specification
+
+This RFC proposes a preference variable to configure the elevation of errors produced by native
+commands to first-class PowerShell errors, so that native command failures will produce error
+objects that are added to the error stream and may terminate execution of the script without added
+boilerplate.
+
+The specification proposes similar functionality to the common POSIX shell configuration `set -eo pipefail`.
+
+- `set -u` - returns an error if any variable has not been previously defined.
```

Comment:
this isn't really part of the new proposal, right?

Reply:
Correct, I'll add a sentence that says this was added for clarity.